Add deep nesting feature

cvle opened the pull request:
Add support for deep nesting.
Fixes https://github.com/cssinjs/jss/issues/312.

kof commented on src/index.js:
```diff
@@ -46,6 +47,35 @@ export default function jssNested() {
 
+  function resolveSelector(parentSelector, nestedSelector, ref) {
+    let result = ''
+    const parentSelectorList = parentSelector.split(separatorRegExp)
```
I don't quite understand why do we split the parent selector, can you give me an example?

cvle replied:
This is an example in SASS but follows the same principle:

```css
  button {
    color: black;
    & .a, .b {
      color: red;
      & .c, &:hover {
        color: gold;
      }
    }
  }
```

becomes
```css
button {
  color: black;
}
button .a, button .b {
  color: red;
}
button .a .c, button .a:hover, button .b .c, button .b:hover {
  color: gold;
}
```

You can see that if the parent selector is a list, you need to apply every nested selector to each selector in the parent selector list.

kof replied:
But once we create the `& .a, .b {` rule, plugin will be called on it so on the next loop we will get the ` & .c, &:hover` rule. We only need to handle one layer as far as I know.

cvle replied:
On first iteration we handle `button` with `& .a, .b` so we create the rule `button .a, button .b`. The plugin will be called on it on the next loop. This time we have `button .a, button .b` and `& .c, &:hover`. To get the desired result `button .a .c, button .a:hover, button .b .c, button .b:hover` we apply each selector from the parent one with the nested ones, thus the need to split both parent and nested selectors. Or did I miss something?

cvle replied:
The only way I see it doing with only _one layer_ is to split each layer in separate rules on each iteration. This would create rules exponentially with the same contents. I did this on my first try and it seemed undesirable.

cvle commented on readme.md:
````diff
@@ -117,13 +117,29 @@ const sheet = jss.createStyleSheet({
+```javascript
+const sheet = jss.createStyleSheet({
+  button: {
+    '&.special': {
````
How about using a ref like `$special`? I'm not sure about using a tag as a selector either.

kof replied:
Hmm, yeah I think $ref example is somewhat better, using global selectors should be probably a separate example for globals.

cvle commented on src/index.js:
```diff
@@ -1,5 +1,6 @@
 import warning from 'warning'
 
+const separatorRegExp = /\s*,\s*/g
```
Btw. `npm run bench` seems to be broken.

cvle replied:
I'll change to normal `split + trim` for better performance.

kof replied:
I have done a test, your regexp is actually 50% faster in chrome https://jsperf.com/split-and-trim-fn-vs-regex

kof replied:
for some reason in firefox its much slower, however I think we might be trapped by some optimizations anyways.

cvle replied:
Oh great your test is a lot more representable than the hacky one I did yesterday. JsPerf was not working for me yesterday. `split+trim` runs really slow on safari at 80% slower. Given that mobile users don't run firefox, I think using the regexp split is the better choice. Anyway this will probably not have much of an impact of the overall performance of any app using this, but given the result of the bench, I'll switch back to regexp.

cvle commented:
I pushed a new revision incorporating suggestions from the review. I left example unchanged as this was still undecided.

kof commented on src/index.js:
```diff
@@ -46,6 +46,28 @@ export default function jssNested() {
 
+  function resolveSelector(parentSelector, nestedSelector, ref) {
+    let result = ''
+    const parentSelectorList = parentSelector.split(',').map((s) => s.trim())
```
you can put  `(s) => s.trim()` as a separate function in the parent scope, so that you don't generate it twice over and over.

cvle replied:
true!

kof commented on src/index.js:
```diff
@@ -46,6 +47,35 @@ export default function jssNested() {
+    const parentSelectorList = parentSelector.split(separatorRegExp)
+    const nestedSelectorList = nestedSelector.split(separatorRegExp)
+
+    for (parentSelector of parentSelectorList) {
```
From what I [can see](https://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=false&presets=es2015%2Ces2015-loose%2Ces2016%2Ces2017%2Clatest%2Creact%2Cstage-0%2Cstage-1%2Cstage-2%2Cstage-3&experimental=false&loose=false&spec=false&code=for%20(const%20parentSelector%20of%20parentSelectorList)%20%7B%0A%7D), the transpiled version is an overkill for this plugin

cvle replied:
Interesting.. I never took that close of a look.

cvle commented:
New revision:
  - Changed for..of to normal for loops
  - Switched back to regexp split
  - Use ref in example and changed `special` to `warn` for a more concrete example.

kof commented:
released!
